api/envoy/extensions/filters/http/credential_injector/v3/credential_injector.proto

```diff
@@ -0,0 +1,113 @@
+syntax = "proto3";
+
+package envoy.extensions.filters.http.credential_injector.v3;
+
+import "envoy/config/core/v3/http_uri.proto";
+import "envoy/extensions/transport_sockets/tls/v3/secret.proto";
+
+import "xds/annotations/v3/status.proto";
+
+import "udpa/annotations/status.proto";
+import "validate/validate.proto";
+
+option java_package = "io.envoyproxy.envoy.extensions.filters.http.credential_injector.v3";
+option java_outer_classname = "CredentialInjectorProto";
+option java_multiple_files = true;
+option go_package = "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/http/credential_injector/v3;credential_injectorv3";
+option (udpa.annotations.file_status).package_version_status = ACTIVE;
+option (xds.annotations.v3.file_status).work_in_progress = true;
+
+// [#protodoc-title: Credential Injector]
+// [#not-implemented-hide:]
+// Credential Injector :ref:`configuration overview <config_http_filters_credential_injector>`.
+// [#extension: envoy.filters.http.credential_injector]
+
+// Filter config.
+message CredentialInjector {
+  // Credential to inject into the proxied requests.
+  Credential credential = 1 [(validate.rules).message = {required: true}];
```

Comment:
> Alternatively to having a `Credential` type with `oneof` inside it, it might be better to declare `credential` field of type `TypedExtensionConfig`, e.g.
> 
> ```proto
> message CredentialInjector {
>   // Credential to inject into the proxied requests.
>   core.v3.TypedExtensionConfig credential = 1 [(validate.rules).message = {required: true}];
> }
> ```
> 
> `core.v3.TypedExtensionConfig` is used in other Envoy APIs:
> 
> * [AuditLoggerConfig](https://github.com/envoyproxy/envoy/blob/37f764e056505632ae12a22d215fa46305be4053/api/envoy/config/rbac/v3/rbac.proto#L119)
> * [OnMatch](https://github.com/envoyproxy/envoy/blob/37f764e056505632ae12a22d215fa46305be4053/api/envoy/config/common/matcher/v3/matcher.proto#L43)
> * [ClusterSpecifierPlugin](https://github.com/envoyproxy/envoy/blob/37f764e056505632ae12a22d215fa46305be4053/api/envoy/config/route/v3/route_components.proto#L511)

@yskopets that's an interesting point!

@lizan @kyessenov is there any guideline on when to use ```one of```  and when to use ```TypedExtensionConfig```?

It seems to me that if the options are open-ended and there are many potential candidates, ```TypedExtensionConfig``` should be used to provide flexibility and avoid a lengthy ```one of``` struct. If the options are limited and unlikely to expand significantly, using ```one of``` may be a simpler and more straightforward solution.

Reply:
The question would be whether the number of types of credentials would grow. I slightly prefer it to be extension so it is more future-proof.

Reply:
It's very likely the options would grow since there are other workload authentication mechanisms we may want to support, such as [OAuth2 resource owner password credentials grant ](https://www.rfc-editor.org/rfc/rfc6749#section-4.3), [STS](https://en.wikipedia.org/wiki/Security_token_service).

Reply:
@yskopets, @lizan it seems that each category should be implemented as an independent extension with its own factory method. Additionally, we would need to define a "CredentialInjector" interface that will be used by all possible extensions. However, considering that the processing for each category may vary significantly, defining a common interface might be challenging. Additionally, all existing HTTP filters do not utilize this two-level extension approach; it is primarily used for platform-level extensions such as logger and cluster extensions.

Given the above reasons, should we still use ```one of``` for credential types?

Reply:
There are multiple benefits of having typed-extensions (e.g., selecting which extensions will be built as part of the binary, allowing additional extensions without changing the current proto, etc.).

Can you please explain why defining a common interface might be challenging?

Reply:
Because there are many kinds of client credentials, and they may be obtained through  different ways.  In the simplest case, we could just pass the HTTP headers to this ```common interface```, but it could also be as complex as the user auth code flow, which needs multiple interaction with the downstreams.

The other reason I'm not in favor of typed-extensions is that this filter itself will be very thin, may be just a couple of lines - only invoking the different extensions.  There's not much common logic here.

Reply:
@zhaohuabing Feel free to go ahead with your original proposal

Reply:
The downside I can see to this approach is that it may use multiple other extensions, and then using the credential-injector filter will require pulling in other extensions. Having typed-extensions will allow selective addition of the other extensions as needed.
That said, as the extension is defined as WIP, it may be changed later (not desirable, but allowed).
@lizan may have other comments